1. Problem 1.1: Grade Calculator (Easy)
Write a program that takes a student's marks (0-100) as input and prints their grade according to the following criteria:

90-100: Grade A
80-89: Grade B
70-79: Grade C
60-69: Grade D
Below 60: Grade F
Also, validate that the marks are within the valid range (0-100).

Hint: Use if-elif-else statements. First check if marks are in valid range (0-100), then use >= comparisons starting from highest grade.

In [ ]:
student_marks = int(input())

if student_marks  < 0 or student_marks > 100:
  print("Invalid Range")
elif student_marks >= 90:
    print("Grade A")
elif student_marks >= 80:
    print("Grade B")
elif student_marks >= 70:
    print("Grade C")
elif student_marks >= 60:
    print("Grade D")
elif student_marks >= 0:
    print("Grade F")


75
Grade C


Problem 1.2: Leap Year Checker (Medium)
Write a program to check if a given year is a leap year or not.

Rules:

A year is a leap year if it is divisible by 4
However, if it is divisible by 100, it is NOT a leap year
Exception: If it is also divisible by 400, then it IS a leap year
Examples:

2000: Leap year (divisible by 400)
1900: Not a leap year (divisible by 100 but not 400)
2024: Leap year (divisible by 4 but not 100)
Hint: Check conditions in order: first divisible by 400, then by 100, then by 4. Use modulo operator (%) to check divisibility.

In [ ]:
year = int(input())

if year % 400 == 0:
  print("Leap Year")

elif year % 100 == 0:
  print("Not Leap Year")

elif year % 4 == 0:
  print("Leap Year")

else:
  print("Not Leap year")


1900
Not Leap Year


Problem 1.3: Triangle Type Identifier (Medium)
Write a program that takes three sides of a triangle as input and determines:

Whether the three sides can form a valid triangle (sum of any two sides must be greater than the third side)
If valid, classify the triangle as:
Equilateral (all three sides equal)
Isosceles (exactly two sides equal)
Scalene (no sides equal)
Hint: First check: (a+b>c) and (b+c>a) and (a+c>b). Then use == to compare sides: if all three equal, it's equilateral; if any two equal, it's isosceles; otherwise scalene.

In [ ]:
a = 10
b = 8
c = 15

if (a + b) > c and (b + c) > a and (c + a) > b:
  if (a == b and b == c):
    print("Equilateral")
  elif (a == b) or (b == c) or (a == c):
    print("Isosceles")
  elif (a != b) and (b != c) and (a != c):
    print("Scalene")

Scalene


Problem 1.4: BMI Calculator with Categories (Easy)
Write a program that calculates BMI (Body Mass Index) and categorizes it.

Formula: BMI = weight (kg) / (height (m))²

Categories:

Below 18.5: Underweight
18.5-24.9: Normal weight
25-29.9: Overweight
30 and above: Obese
Hint: First calculate BMI using the formula (use ** for power or multiply height by itself). Then use if-elif-else to categorize.

In [ ]:
weight = 55
height = 1.65

result = weight / (height * height)
print(result)

if result < 18.5:
  print("Underweight")
elif result <= 24.9:
  print("Normal weight")
elif result <= 29.9:
  print("Overweight")
else :
  print("Obese")


20.202020202020204
Normal weight


Section 2: For Loops
Problem 2.1: Multiplication Table (Easy)
Write a program that prints the multiplication table for a given number (1 to 10).

Example output for number 5:

5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
...
5 x 10 = 50
Hint: Use for loop with range(1, 11) to iterate from 1 to 10. Multiply the number by loop variable i.

In [ ]:
a = int(input())

for i in range(1,11):
  print(f"{a}  *  {i}  = {a * i}")


5  *  1  = 5
5  *  2  = 10
5  *  3  = 15
5  *  4  = 20
5  *  5  = 25
5  *  6  = 30
5  *  7  = 35
5  *  8  = 40
5  *  9  = 45
5  *  10  = 50


Problem 2.2: Factorial Calculator (Medium)
Write a program to calculate the factorial of a number using a for loop.

Factorial of n (n!) = n × (n-1) × (n-2) × ... × 1

Example: 5! = 5 × 4 × 3 × 2 × 1 = 120

Hint: Initialize result = 1. Use for loop with range(1, n+1) and multiply result by each number.

In [ ]:
a = int(input())

result = 1

for i in range(1,a+1):
  result = result * i
print(result)

120


Problem 2.3: Sum of Even Numbers (Easy)
Write a program that calculates the sum of all even numbers between 1 and n (inclusive), where n is provided by the user.

Example: If n = 10, sum = 2 + 4 + 6 + 8 + 10 = 30

Hint: Use for loop with range(1, n+1). Check if i % 2 == 0, then add to sum. Or use range(2, n+1, 2) to iterate only even numbers.

In [ ]:
a = int(input())
result = 0
for i in range(2,a+1,2):
  result = result + i
print(result)


30


In [ ]:
a = int(input())

result = 0
for i in range(1,a+1):
  if i%2 == 0:
    result = result + i
print(result)

30


Problem 2.4: Fibonacci Series (Medium)
Write a program to print the first n terms of the Fibonacci series.

The Fibonacci series: 0, 1, 1, 2, 3, 5, 8, 13, 21, ...

Each number is the sum of the two preceding ones.

Hint: Initialize a=0, b=1. Print a, then in loop: calculate next=a+b, print next, then update a=b and b=next.

In [ ]:
a = int(input())

result = [0,1]

for i in range(2, a):
  final = (result[-2] + result[-1])
  result.append(final)
print(result)



5
[0, 1, 1, 2, 3]


Problem 2.5: Prime Number Checker (Medium)
Write a program to check if a given number is prime or not.

A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself.

Hint: Check if n is divisible by any number from 2 to n-1. If no divisor found, it's prime. Optimize: check only up to √n.

In [ ]:
a = int(input())

if a <= 1:
  print("Not a Prime")
else:
  for i in range(2, a-1):
    if (a % i ==0):
      print("Not a Prime")
      break
  else:
    print("Prime")




5
Prime


Problem 2.6: Count Digits in a Number (Easy)
Write a program that counts the number of digits in a given integer.

Example: 12345 has 5 digits

Hint: Convert number to string and use len(), or keep dividing by 10 until number becomes 0, counting each division.

In [ ]:
a = int(input())
b = str(a)

result = len(b)
print(result)

567893
6


Section 3: While Loops
Problem 3.1: Reverse a Number (Medium)
Write a program that reverses the digits of a given number using a while loop.

Example: Input: 12345, Output: 54321

Hint: Initialize reverse=0. While n>0: get last digit (n%10), add to reverse (reverse=reverse*10+digit), remove last digit (n=n//10).

In [ ]:
a = int(input())

reverse = 0

while a > 0:
  digit = a % 10
  reverse = reverse * 10 + digit
  a = a // 10

print(reverse)



12345
54321


Problem 3.2: Sum of Digits (Easy)
Write a program that calculates the sum of all digits in a given number.

Example: Input: 12345, Output: 15 (1+2+3+4+5)

Hint: While n>0: add last digit to sum (sum += n%10), then remove it (n = n//10).

In [ ]:
a = int(input())

result = 0

while a > 0:
  digit = a % 10
  result = result + digit
  a = a // 10
print(result)


12345
15


Problem 3.3: Number Guessing Game (Medium)
Write a number guessing game where:

The program generates a random number between 1 and 100
The user has to guess the number
After each guess, the program tells the user if their guess is too high or too low
The game continues until the user guesses correctly
Display the number of attempts taken
Hint: Use import random and random.randint(1, 100) to generate number. Use while loop with condition guess != target. Count attempts.


In [ ]:
import random

target = random.randint(1, 100)

attempts = 0

guess = int(input("Guess a number between 1 and 100: "))

while guess != target:
  attempts = attempts + 1

  if guess > target:
    print("Too High")
  else:
    print("Too Low")

  guess = int(input("Guess again: "))

attempts = attempts + 1

print("Correct!")
print("Number of attempts : ", attempts)


Guess a number between 1 and 100: 10
Too Low
Guess again: 20
Too Low
Guess again: 50
Too High
Guess again: 40
Too Low
Guess again: 45
Too Low
Guess again: 48
Too Low
Guess again: 49
Correct!
Number of attempts :  7


Problem 3.4: Palindrome Checker (Medium)
Write a program to check if a given number is a palindrome or not.

A palindrome number reads the same forwards and backwards.

Examples: 121, 12321, 1331 are palindromes

Hint: Store original number, reverse it (like Problem 3.1), then compare original with reversed number.

In [ ]:
a = int(input())

original = a

reverse = 0

while a > 0:
  digit = a % 10
  reverse = reverse * 10 + digit
  a = a // 10
print(reverse)

if original == reverse:
  print("Palindrome")
else:
  print("Not a Palindrome")

121
121
Palindrome


Problem 3.5: Armstrong Number (Hard)
Write a program to check if a number is an Armstrong number (also called Narcissistic number).

An Armstrong number is a number that is equal to the sum of its own digits each raised to the power of the number of digits.

Examples:

153 = 1³ + 5³ + 3³ = 1 + 125 + 27 = 153 (Armstrong number)
9474 = 9⁴ + 4⁴ + 7⁴ + 4⁴ = 6561 + 256 + 2401 + 256 = 9474 (Armstrong number)
Hint: First count digits (power). Store original. Extract each digit, raise to power, add to sum. Compare sum with original.



In [ ]:
a = int(input())

b = str(a)

length = len(b)
result = 0

for i in str(b):
  final = int(i)
  result = result + (final ** length)

if a == result:
  print("Armstrong number")
else:
  print("Not Armstrong number")



123
Not Armstrong number


Section 4: Nested Loops
Problem 4.1: Pattern Printing - Right Triangle (Easy)
Print the following pattern for n rows:

*
**
***
****
*****
(Example shown for n=5)

Hint: Use nested loops: outer loop for rows (i from 1 to n), inner loop for columns (j from 1 to i). Print * in inner loop.



In [ ]:
a = int(input())

for i in range(1,a+1):
  for j in range(1,i+1):
    print("*", end="")
  print()



5
*
**
***
****
*****


In [ ]:
a = int(input())

for i in range(1,a+1):
  print("*" * i)

5
*
**
***
****
*****


Problem 4.2: Pattern Printing - Inverted Right Triangle (Easy)
Print the following pattern for n rows:

*****
****
***
**
*
(Example shown for n=5)

Hint: Outer loop for rows. Inner loop prints stars from 1 to (n-i+1), where i is the row number.



In [ ]:
a = int(input())

for i in range(a):
  for j in range(a - i):
    print("*", end="")
  print()

5
*****
****
***
**
*


In [ ]:
a = int(input())

for i in range(a):

  b = (a-i)
  print("*" * b)

5
*****
****
***
**
*


Problem 4.3: Pattern Printing - Pyramid (Medium)
Print the following pyramid pattern for n rows:

    *
   ***
  *****
 *******
*********
(Example shown for n=5)

Hint: For row i: print (n-i) spaces, then print (2*i-1) stars. Use nested loops or string multiplication.



In [ ]:
a = int(input())

for i in range(1,a+1):
  final = (" " * (a - i)) + ("*" * (2 * i - 1))
  print(final)

6
     *
    ***
   *****
  *******
 *********
***********


Problem 4.4: Pattern Printing - Number Pattern (Medium)
Print the following number pattern for n rows:

1
12
123
1234
12345
(Example shown for n=5)

Hint: Similar to Problem 4.1, but print j (column number) instead of *.

In [ ]:
a = int(input())

for i in range(1,a+1):
  for j in range(1,i+1):
    print(j, end="")
  print()

6
1
12
123
1234
12345
123456


Problem 4.5: Multiplication Table Grid (Medium)
Print a multiplication table grid from 1 to n.

Example for n=5:

1   2   3   4   5
2   4   6   8  10
3   6   9  12  15
4   8  12  16  20
5  10  15  20  25
Hint: Nested loops: for i in range(1, n+1) and for j in range(1, n+1). Print i*j with proper spacing.



In [ ]:
a = int(input())

for i in range(1,a+1):
  for j in range(1,a+1):
    print(i*j,end=" ")
  print()

6
1 2 3 4 5 6 
2 4 6 8 10 12 
3 6 9 12 15 18 
4 8 12 16 20 24 
5 10 15 20 25 30 
6 12 18 24 30 36 


Problem 4.6: Diamond Pattern (Hard)
Print a diamond pattern for n rows (n should be odd):

    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *
(Example shown for n=9)

Hint: Print pyramid (upper half) for rows 1 to (n+1)//2, then print inverted pyramid (lower half) for remaining rows.



In [ ]:
a = 5

for i in range(1,a+1):
  result = " " * (a-i) + ("*" * (2*i -  1))
  print(result)
  result = " " *

    *
   ***
  *****
 *******
*********


In [ ]:
a = int(input())

b = (a+1) //2

for i in range(1,b+1):
  print(" " * (b-i) + "*" * (2*i -1))

for i in range(b-1,0,-1): ### start,stop,step
  print(" " * (b-i) + "*" * (2*i -1))


5
  *
 ***
*****
 ***
  *
